### PR TITLE
[Add] Integrated semantic search

### PR DESCRIPTION
**Pull request by wangzelin007**
1. Having high-quality answers
```json
{
    "data": [
        {
            "description": "Create a managed disk from a snapshot",
            "firstCommand": "az snapshot show",
            "source": 1,
            "commandSet": [
                {
                    "command": "az snapshot show",
                    "arguments": [
                        "--name",
                        "--resource-group",
                        "--query",
                        "-o"
                    ],
                    "reason": "Get the snapshot Id (snapshotId will be used in subsequent commands).",
                    "example": "az snapshot show --name $snapshotName --resource-group $resourceGroupName --query [id] -o tsv"
                },
                {
                    "command": "az disk create",
                    "arguments": [
                        "--resource-group",
                        "--name",
                        "--sku",
                        "--size-gb",
                        "--source"
                    ],
                    "reason": "Create a new Managed Disks using the snapshot Id\\nNote that managed disk will be created in the same location as the snapshot",
                    "example": "az disk create --resource-group $resourceGroupName --name $diskName --sku $storageType --size-gb $diskSize --source $snapshotId"
                }
            ],
            "id": "Create a managed disk from a snapshot with CLI (Linux)",
            "source_url": "https://docs.microsoft.com/en-us/azure/virtual-machines/scripts/create-managed-disk-from-snapshot#run-the-script",
            "update_time": "2022-08-23T06:14:20.366Z",
            "@search.score": 8.918325,
            "@search.highlights": {
                "description": [
                    "<em>Create</em> a <em>managed</em> disk from a snapshot"
                ],
                "name": [
                    "<em>Create</em> a <em>managed</em> disk from a snapshot with CLI (Linux)"
                ]
            },
            "score": 3.338235378265381,
            "highlights": {
                "scenario": "<em>Create</em> a<em> managed disk</em> from a<em> snapshot</em> with CLI (Linux).<em> Create</em> a<em> managed disk</em> from a<em> snapshot.</em>"
            },
            "scenario": "Create a managed disk from a snapshot with CLI (Linux)"
        }
    ],
    "error": null,
    "status": 200,
    "api_version": "1.0.0"
}
```
2. No high-quality answers or no answer
```
{"data": [], "error": null, "status": 200, "api_version": "1.0.0"}
```

### CODE REVIEW

File: API/common/service_impl/knowledge.py [hunk 1]
```diff
@@ -28,44 +28,87 @@ class MatchRule(int, Enum):
...
     results = list(results)
     for result in results:
         result.pop("rid")
-        result["score"] = result.pop("@search.score")
-        result["highlights"] = result.pop("@search.highlights")
+        if search_type == SearchType.Semantic:
+            result["score"] = result.pop("@search.reranker_score")
+            captions = result.pop("@search.captions")
+            if captions:        
+                result["highlights"] = {"scenario": captions[0].highlights}
```
Comment (zhoxing-ms) [inline API/common/service_impl/knowledge.py]: Just out of curiosity, may I ask why we only take the highlights from the first item of `captions`?

File: API/CopilotService/__init__.py [hunk 2]
```diff
@@ -25,8 +25,14 @@ def main(req: func.HttpRequest) -> func.HttpResponse:
...
     try:
         result = []
+        error = None
         if service_type == ServiceType.KNOWLEDGE_SEARCH:
             result = knowledge_search(question, top_num)
+            if len(result) > 0 and result[0]['score'] < float(os.environ.get('KNOWLEDGE_QUALITY_THRESHOLD', "1.0")):
+                error = f"Knowledge quality is too low, score: {result[0]['score']}"
+                result = []
+            elif len(result) == 0:
+                error = "No knowledge found"
```
Comment (zhoxing-ms) [inline API/CopilotService/__init__.py]: In fact, I am not sure if the reason of none high-quality data should be defined as error message.
If the caller doesn't care about the specific details of the reason for none data, it's just that we need to know this information when troubleshooting, could we replace it with printing log?

File: API/common/service_impl/knowledge.py [hunk 3]
```diff
@@ -28,44 +28,87 @@ class MatchRule(int, Enum):
...
+class SearchType(int, Enum):
+    Semantic = 1
+    FullText = 2
+    Keyword = 3
+
+def knowledge_search(keyword: str, top_num: int):
+    # placeholder for rate limit
+    rate_limit = False
+    if not rate_limit:
+        results = knowledge_search_semantic(keyword, top_num)
```
Comment (zhoxing-ms) [inline API/common/service_impl/knowledge.py]:
```suggestion
    exceed_rate_limit = False
    if not exceed_rate_limit:
        results = knowledge_search_semantic(keyword, top_num)
```

File: API/common/service_impl/knowledge.py (hunk 1 above)
Comment (wangzelin007) [reply API/common/service_impl/knowledge.py]: During testing, I found that the results of semantic search highlighting is a list. (The highlighted results of full-text search is a dictionary)
And there seems to be no case where the list length > 1, so I just took the first item.
To be on the safe side, add logic to process the entire list, separated by `;`

File: API/CopilotService/__init__.py [hunk 4]
```diff
@@ -27,6 +27,11 @@ def main(req: func.HttpRequest) -> func.HttpResponse:
         result = []
         if service_type == ServiceType.KNOWLEDGE_SEARCH:
             result = knowledge_search(question, top_num)
+            if len(result) > 0 and result[0]['score'] < float(os.environ.get('KNOWLEDGE_QUALITY_THRESHOLD', "1.0")):
+                print(f"Knowledge quality is too low, score: {result[0]['score']}")
```
Comment (zhoxing-ms) [inline API/CopilotService/__init__.py]: Please use `logging.info` *(need to import `logging`)* instead of `print`

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**